grpc/swift-cubic/Sources/swift-cubic/CubicManager.swift

```diff
@@ -0,0 +1,168 @@
+                                                           connectionBackoff: nil)
+        let connection = ClientConnection.init(configuration: configuration)
+        self.client = Cobaltspeech_Cubic_CubicServiceClient(connection: connection)
+    }
+    
+    public init(host: String,
+                port: Int,
+                tlsCertificateFileName: String,
+                tlsCertificateFormat: NIOSSLSerializationFormats) {
+        let target = ConnectionTarget.hostAndPort(host, port)
+        let eventLoopGroup = PlatformSupport.makeEventLoopGroup(loopCount: 1,
+                                                                networkPreference: .best)
+        
+        var configuration = ClientConnection.Configuration(target: target, eventLoopGroup: eventLoopGroup)
+        
+        if let cert = try? NIOSSLCertificate(file: tlsCertificateFileName, format: tlsCertificateFormat) {
+            let source = NIOSSLCertificateSource.certificate(cert)
+            var tls = ClientConnection.Configuration.TLS()
+            tls.certificateChain.append(source)
+            configuration.tls = tls
+        }
+        
+        let connection = ClientConnection.init(configuration: configuration)
+        self.client = Cobaltspeech_Cubic_CubicServiceClient(connection: connection)
+    }
+    
+    public func listModels(success: @escaping (_ models: [Cobaltspeech_Cubic_Model]?) -> (), failure: CubicFailureCallback?) {
+        let request = Cobaltspeech_Cubic_ListModelsRequest()
+        
+        client.listModels(request).response.whenComplete({ (result) in
+            do {
+                let response = try result.get()
+                success(response.models)
+            } catch let error {
+                failure?(error)
+            }
+        })
+    }
+    
+    public func sendStreamPartToServer(_ audioData: Data, failure: CubicFailureCallback?) {
+        guard let callStream = callStream else { return }
+        
+        var request = Cobaltspeech_Cubic_StreamingRecognizeRequest()
+        request.audio.data = audioData
+        
+        callStream.sendMessage(request).whenComplete({ (result) in
+            switch result {
+            case .failure(let error):
+                failure?(error)
+            default:
+                break
+            }
+        })
+    }
+    
+    public func startStream(success: @escaping (_ response: Cobaltspeech_Cubic_RecognitionResponse) -> (), failure: CubicFailureCallback?) {
+        guard let selectedModel = selectedModel else { return }
+        
+        let call = self.client.streamingRecognize(handler: { (result) in
+            success(result)
+        })
+
+        var request = Cobaltspeech_Cubic_StreamingRecognizeRequest()
+        request.config.modelID = selectedModel.id
+
+        if let idleTimeoutSeconds = idleTimeoutSeconds {
+            request.config.idleTimeout.seconds = idleTimeoutSeconds
+        }
+        
+        if let audioEncoding = audioEncoding {
```

Comment:
Pardon my swift ignorance, but what `if let x = x`  do?

Reply:
When x is has an optional type (nullable), often we need to safely unwrap it (check if it != nil and then use it as not-nullable variable).

`if let x = x {
    y = x
}` means:

`if x != nil {
    y = x!
}`

We cannot simply write `y = x`, if y is non-optional and x is optional. And if x is nil, we shouldn't set y in order to leave its default value.
So this is normal and generally accepted way to use `if let x = x`, when we need to exclude cases when x is nil.

Reply:
got it! thanks!